static/app/utils/profiling/flamegraph.ts

```diff
@@ -351,4 +357,70 @@ export class Flamegraph {
 
     return matches;
   }
+
+  collapseSystemFrames() {
+    const {root, frames, depth} = buildCollapsedFlamegraph(collapse(this.root));
+    this.root = root;
+    this.depth = depth;
+    this.frames = frames;
+  }
+}
+
+const isSystemFrame = (frame: FlamegraphFrame) => !frame.node.frame.is_application;
+
+function collapse(frame: FlamegraphFrame) {
+  frame.children = frame.children.reduce((acc, child) => {
+    acc.push(collapse(child));
+    return acc;
+  }, [] as FlamegraphFrame[]);
+
+  if (!frame.parent) {
+    return frame;
+  }
+
+  const children = frame.children;
+
+  if (children.length === 1) {
```

Comment:
Why do you care if there is only one child?

Reply:
its to visually preserve branching, although we could change this to collapse all children if they're all are system frames but you might lose context of what's calling the application function